![New York City schoolbus](schoolbus.jpg)

Photo by [Jannis Lucas](https://unsplash.com/@jannis_lucas) on [Unsplash](https://unsplash.com).
<br>

Every year, American high school students take SATs, which are standardized tests intended to measure literacy, numeracy, and writing skills. There are three sections - reading, math, and writing, each with a **maximum score of 800 points**. These tests are extremely important for students and colleges, as they play a pivotal role in the admissions process.

Analyzing the performance of schools is important for a variety of stakeholders, including policy and education professionals, researchers, government, and even parents considering which school their children should attend. 

You have been provided with a dataset called `schools.csv`, which is previewed below.

You have been tasked with answering three key questions about New York City (NYC) public school SAT performance.

In [101]:
# Re-run this cell 
import pandas as pd

# Read in the data
schools = pd.read_csv("schools.csv")

# Preview the data
schools.head()


school_name  ... percent_tested
0  New Explorations into Science, Technology and ...  ...            NaN
1                               Essex Street Academy  ...           78.9
2                       Lower Manhattan Arts Academy  ...           65.1
3    High School for Dual Language and Asian Studies  ...           95.9
4      Henry Street School for International Studies  ...           59.7

[5 rows x 7 columns]

In [102]:
# Display number of rows
len(schools)

375

In [103]:
# Check for null values
schools.isna().sum()

school_name         0
borough             0
building_code       0
average_math        0
average_reading     0
average_writing     0
percent_tested     20
dtype: int64

In [104]:
# Print rows with null values
schools[schools["percent_tested"].isnull()]

school_name  ... percent_tested
0    New Explorations into Science, Technology and ...  ...            NaN
19   Urban Assembly School of Business for Young Women  ...            NaN
38               High School of Hospitality Management  ...            NaN
57      Frederick Douglass Academy II Secondary School  ...            NaN
76              Jacqueline Kennedy Onassis High School  ...            NaN
95                             Tottenville High School  ...            NaN
114                    Bronx International High School  ...            NaN
133                  West Bronx Academy for the Future  ...            NaN
152              Bronx High School for the Visual Arts  ...            NaN
171              Kingsbridge International High School  ...            NaN
190  Archimedes Academy for Math, Science, and Tech...  ...            NaN
209         Young Women's Leadership School in Astoria  ...            NaN
228  High School for Public Service: Heroes of Tomo...  ...            NaN
247                          Multicultural High School  ...            NaN
266                         Boys and Girls High School  ...            NaN
285                 Brooklyn Academy of Global Finance  ...            NaN
304  Brooklyn Secondary School for Collaborative St...  ...            NaN
323                               Flushing High School  ...            NaN
342             Humanities and Arts Magnet High School  ...            NaN
361                              Hillcrest High School  ...            NaN

[20 rows x 7 columns]

In [105]:
# Subset columns school_name and average_math
best_math_schools = schools[["school_name","average_math"]]

In [106]:
# Subset rows where average math score is at least 640
best_math_schools = best_math_schools[best_math_schools["average_math"]>=640]

In [107]:
# Sort values in descending order
best_math_schools = best_math_schools.sort_values('average_math',ascending=False)
best_math_schools

school_name  average_math
88                              Stuyvesant High School           754
170                       Bronx High School of Science           714
93                 Staten Island Technical High School           711
365  Queens High School for the Sciences at York Co...           701
68   High School for Mathematics, Science, and Engi...           683
280                     Brooklyn Technical High School           682
333                        Townsend Harris High School           680
174  High School of American Studies at Lehman College           669
0    New Explorations into Science, Technology and ...           657
45                       Eleanor Roosevelt High School           641

Stuyvesant High School has the best math results, with the average math score of 754.

In [108]:
# Add new column for total SAT scores
schools["total_SAT"] = schools["average_math"] + schools["average_reading"] + schools["average_writing"]

In [109]:
# Save results in dataframe
top_10_schools = schools[["school_name","total_SAT"]]

top_10_schools = top_10_schools.sort_values("total_SAT", ascending=False).head(10)
top_10_schools

school_name  total_SAT
88                              Stuyvesant High School       2144
170                       Bronx High School of Science       2041
93                 Staten Island Technical High School       2041
174  High School of American Studies at Lehman College       2013
333                        Townsend Harris High School       1981
365  Queens High School for the Sciences at York Co...       1947
5                       Bard High School Early College       1914
280                     Brooklyn Technical High School       1896
45                       Eleanor Roosevelt High School       1889
68   High School for Mathematics, Science, and Engi...       1889

The top 10 performing schools are as below :-
1.  Stuyvesant High School
2.  Bronx High School of Science
3.  Staten Island Technical High School
4.  High School of American Studies at Lehman College
5.  Townsend Harris High School
6.  Queens High School for the Sciences at York College
7.  Bard High School Early College
8.  Brooklyn Technical High School
9.  Eleanor Roosevelt High School
10. High School for Mathematics, Science, and Engineering at City College

In [110]:
# Average and standard deviation of total SAT scores for each borough
std_mean_school = schools.groupby("borough")["total_SAT"].agg(['count', 'mean', 'std']).round(2)
std_mean_school

count     mean     std
borough                              
Bronx             98  1202.72  150.39
Brooklyn         109  1230.26  154.87
Manhattan         89  1340.13  230.29
Queens            69  1345.48  195.25
Staten Island     10  1439.00  222.30

In [111]:
# Select the borough with the largest standard deviation
largest_std_dev = std_mean_school[std_mean_school["std"] == std_mean_school["std"].max()]
largest_std_dev

count     mean     std
borough                          
Manhattan     89  1340.13  230.29

In [112]:
# Rename the columns in the dataframe
largest_std_dev = largest_std_dev.rename(columns={"count":"num_schools", "mean":"average_SAT", "std":"std_SAT"})
largest_std_dev

num_schools  average_SAT  std_SAT
borough                                     
Manhattan           89      1340.13   230.29

Manhattan has the largest standard deviation in the combined SAT score.